hnswlib/hnswalg.h

```diff
@@ -451,6 +451,10 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
         tableint next_closest_entry_point = selectedNeighbors.back();
 
         {
+            std::unique_lock <std::mutex> lock(link_list_locks_[cur_c], std::defer_lock);
```

Comment:
I wonder, what is the purpose of `std::defer_lock`? Are there deadlocks without it?

Reply:
without defer_lock we have a dead lock in `tableint addPoint(...)` method because there `mutuallyConnectNewElement` is called under the locking the `link_list_locks_[cur_c])`. So we have two locks of `cur_c`.
I lock the `cur_c` only if element is updated. it helps to avoid dead lock when element is added.